```{r}
setwd("evaluating_coregulation")
source("../downstreamer_functions.r")
source("data/source_olivier.r")

# Read reference datasets
ensembl <- read.table(ensembl.file,sep="\t", header=T, row.names = 1, stringsAsFactors = F)
ensembl$gene.length = ensembl$Gene.end..bp. - ensembl$Gene.start..bp.
rownames(ensembl) <- make.names(ensembl$Gene.name, unique=T)

# Coregulation
files    <- list.files(main.downstreamer.output.path, pattern="*\\_enrichtments\\_exHla\\.xlsx", full.names = T)
datasets <- list()
for (file in files) {
  name <- gsub("\\_hg19\\_enrichtments\\_exHla\\.xlsx", "", basename(file))
  name <- gsub("\\_hg19\\_enrichtments\\_exHla\\_1\\.xlsx", "", name)
  name <- gsub("\\_enrichtments\\_exHla\\.xlsx", "", name)
  name <- gsub("\\_hg19\\.txt\\_exHla\\.xlsx", "", name)
  
  if (length(grep("v55", file)) > 0) {
    name <- paste0(name, "_v55")
  }
  datasets[[name]] <- read.depict2(file)
}

# Genepvalues
files <- list.files(main.downstreamer.output.path, pattern="*_genePvalues_.*.xlsx", full.names = T)
genep <- read.genep.excel(files)
genep[is.na(genep)] <- 1
```

# Distance from SNP vs coreg zscore
```{r}

zscores <- make.zscore.matrix(datasets)


distance.per.gene <- lapply(names(datasets), function(dataset.name){
  dataset <- datasets[[dataset.name]]
  tmp <- dataset$Coregulation
  tmp <- tmp[tmp$Bonferroni.significant==1 & tmp$Enrichment.Z.score > 0,]
  x   <- tmp$Distance.to.indep.GWAS.hit  
  x[x==">1mb"] <- 1000000
  x <- as.numeric(x)
  
  y <- x
  y[x == 0] <- "Gene body"
  y[x > 0 & x < 10000] <- "10k"
  y[x > 10000 & x < 50000] <- "50k"
  y[x > 50000 & x < 250000] <- "250k"
  y[x > 250000] <- ">500k"

  names(y) <- tmp$Gene.set
  
  y <- factor(y, levels=c("Gene body", "10k", "50k", "250k", ">500k"))
  return(list(y, zscores[names(y), dataset.name]))
})


names(distance.per.gene) <- names(datasets)


x <- unlist(sapply(distance.per.gene, function(x){x[[1]]}))
y <- unlist(sapply(distance.per.gene, function(x){x[[2]]}))

boxplot(y ~ x)


mean.zscore.per.bin <- sapply(datasets, function(dataset){
  tmp <- dataset$Coregulation
  x <- tmp$Distance.to.indep.GWAS.hit  
  x[x==">1mb"] <- 1000000
  x <- as.numeric(x)
  
  y <- tmp$Enrichment.Z.score
  m1 <- mean(y[x == 0], na.rm=T)
  m2 <- mean(y[x > 0 & x < 10000], na.rm=T)
  m3 <- mean(y[x > 10000 & x < 50000] , na.rm=T)
  m4 <- mean(y[x > 50000 & x < 250000], na.rm=T)
  m5 <- mean(y[x > 250000], na.rm=T)


  out <- c(m1, m2, m3, m4, m5)
  names(out) <- c("Gene body", "10k", "50k", "250k", ">500k")
  return(out)
})
mean.zscore.per.bin <- t(mean.zscore.per.bin)

boxplot(mean.zscore.per.bin, ylab="Mean Coregulation zscore of genes mapping in winow", xlab="Genomic range compared to GWAS hit")
```
